common/optimiser/Optimiser.java

```diff
@@ -45,30 +53,49 @@ public class Optimiser {
     private MPSolver solver;
     private MPSolverParameters parameters;
     private Status status;
-    private boolean hasSolver;
-    private double objectiveValue;
+    private boolean isConstraintsUpToDate;
+    private Double objectiveValue;
 
     public Optimiser() {
         variables = new ArrayList<>();
         constraints = new HashSet<>();
         objectiveCoefficients = new HashMap<>();
         status = Status.NOT_SOLVED;
-        hasSolver = false;
+        objectiveValue = null;
+        isConstraintsUpToDate = true;
     }
 
     public synchronized Status optimise(long timeLimitMillis) {
-        assert hasSolver;
         if (isOptimal()) return status;
+        assert solutionSatisfiesModel();
+        maySetSolver();
+        setSolutionAsHints();
         solver.setTimeLimit(timeLimitMillis);
         status = Status.of(solver.solve(parameters));
-        recordValues();
+        recordSolverValues();
         if (isOptimal()) releaseSolver();
+        assert solutionSatisfiesModel();
         return status;
     }
 
-    private void recordValues() {
-        objectiveValue = solver.objective().value();
-        if (status != Status.NOT_SOLVED && status != Status.ERROR) variables.forEach(OptimiserVariable::recordSolutionValue);
+    private void maySetSolver() {
+        if (solver == null) createSolver();
+        else if (!isConstraintsUpToDate) {
+            solver.reset();
+            setConstraintCoefficients();
+        }
```

Comment:
We should reset `objectiveValue` to `null` here, otherwise down at line 90 we're comparing with the objective value based on outdated costs.

Reply:
actually not required, because the Constraint coefficients don't affect the objective - the objective is updated only in `setObjectiveCoefficient` and/or nulled out

Reply:
Right, missed that.